File: lab/src/main/java/org/example/web/utils/web/RouteTree.java
package org.example.web.utils.web;

import GlobalEnums.StringEnums;
import org.example.web.RequestHandler;
import org.example.web.utils.Pair;

import java.util.*;

class MatchCandidate {
    private RouteTree node;
    private HashMap<String, String> pathVariables;

    public MatchCandidate(RouteTree node, Map<String, String> pathVariables) {
        this.node = node;
        this.pathVariables = new HashMap<>(pathVariables);
    }

    public RouteTree getNode() {
        return node;
    }

    public void setNode(RouteTree node) {
        this.node = node;
    }

    public HashMap<String, String> getPathVariables() {
        return pathVariables;
    }

    public void setPathVariables(HashMap<String, String> pathVariables) {
        this.pathVariables = pathVariables;
    }
}

public class RouteTree {
//    private final String segment;
    private final boolean isWildcard;
    private final String path;
    private final Map<String, RouteTree> staticRoutes;

    public Map<String, RouteTree> getDynamicRoutes() {
        return dynamicRoutes;
    }

    public String getPath() {
        return path;
    }

    public boolean isWildcard() {
        return isWildcard;
    }

    private final Map<String, RouteTree> dynamicRoutes;


    private RequestHandler requestHandler;
    @Deprecated
    public RouteTree(String segment){
        this.staticRoutes = new HashMap<>();
        this.dynamicRoutes = new HashMap<>();
        this.path = null;
        this.requestHandler = null;
        this.isWildcard = false;
    }
    public RouteTree(String path, boolean wildcard){
        this.staticRoutes = new HashMap<>();
        this.dynamicRoutes = new HashMap<>();
        this.path = path;
        this.requestHandler = null;
        this.isWildcard = wildcard;
    }
    private RouteTree(String segment, String path){
        this.path = path;
        this.staticRoutes = new HashMap<>();
        this.dynamicRoutes = new HashMap<>();
        this.requestHandler = null;
        this.isWildcard = false;
    }
    public static RouteTree parse(String path){
        return  new RouteTree("*",path);
    }

    public Map<String, RouteTree> getStaticRoutes() {
        return staticRoutes;
    }

    public RequestHandler getRequestHandler() {
        return requestHandler;
    }
    public void setRequestHandler(RequestHandler requestHandler) {
        this.requestHandler = requestHandler;
    }
    /**
     * add a route node to current route node
     */
    public void addRoute(String segment, RouteTree route){
        staticRoutes.put(segment, route);
    }

    /**
     * use route-pattern's segment-route, and use current route node as root
     * @param route
     */
    public void addRoute(RoutePattern route){
        RouteTree node = this;
        List<Pair<String, String>> segments = route.getSegments();
        for (int i = 0; i < segments.size(); i++) {
            String segment = segments.get(i).first;
            Map<String, RouteTree> routes = node.staticRoutes;
            if(StringEnums.WILDCARD.getString().equals(segment)){
                routes = node.dynamicRoutes;
                segment = segments.get(i).second;
            }
            if(!routes.containsKey(segment)){
                routes.put(segment, new RouteTree(segment,segments.get(i).second));
            }
            if(i+1 == segments.size()){
                routes.get(segment).setRequestHandler(route.getRequestHandler());
                return;
            }
            node = routes.get(segment);
        }
    }
    public RouteTree find(String path){
        String[] split = path.split("/");
        List<String> segments = Arrays.asList(split);
        Queue<RouteTree> queue = new ArrayDeque<>();
        queue.add(this);
        int idx = 0,length = segments.size();
        while(idx < length && !queue.isEmpty()){
            int size = queue.size();
            for(int i = size; i > 0; --i){
                RouteTree current = queue.poll();
                if( null == current.getStaticRoutes() || current.getStaticRoutes().isEmpty() ){
                    continue;
                }
                RouteTree routeTree = current.getStaticRoutes().get(split[idx]);
                if(routeTree != null){
                    queue.add(routeTree);
                }
                Map<String, RouteTree> children = current.getDynamicRoutes();
                if(children != null  && !children.isEmpty()){
                    queue.addAll(children.values());
                }
            }
            ++idx;
            if(idx == length && !queue.isEmpty()){
                for(int i = queue.size(); i > 0; --i){
                    RouteTree current = queue.poll();
                    if(current.getRequestHandler() != null){
                        return current;
                    }else{
                        queue.add(current);
                    }
                }
            }
        }
        return null;
    }

    public MatchResult findWithVariable(String path){
        String[] split = path.split("/");
        Queue<MatchCandidate> queue = new ArrayDeque<>();
        queue.add(new MatchCandidate(this, new HashMap<>()));
        for (String segment : split) {
            Queue<MatchCandidate> nextLevel = new ArrayDeque<>();
            while (!queue.isEmpty()) {
                MatchCandidate candidate = queue.poll();
                RouteTree routeTree = candidate.getNode().staticRoutes.get(segment);
                if (routeTree != null) {
                    nextLevel.add(new MatchCandidate(routeTree, new HashMap<>(candidate.getPathVariables())));
                }
                Collection<RouteTree> values = candidate.getNode().dynamicRoutes.values();
                if (values != null && !values.isEmpty()) {
                    for (RouteTree value : values) {
                        Map<String, String> pathVariables = new HashMap<>(candidate.getPathVariables());
                        pathVariables.put(value.path, segment);
                        nextLevel.add(new MatchCandidate(value, pathVariables));
                    }
                }
            }
            queue = nextLevel;
        }
        for (MatchCandidate matchCandidate : queue) {
            if(matchCandidate.getNode().getRequestHandler() != null){
                return new MatchResult(matchCandidate.getNode().getRequestHandler(), matchCandidate.getPathVariables());
            }
        }
        return null;
    }

    public void removeRoute(String segment){
        staticRoutes.remove(segment);
    }

    @Override
    public String toString() {
        return "RouteTree{" +
                "isWildcard=" + isWildcard +
                ", path='" + path + '\'' +
                ", staticRoutes=" + staticRoutes +
                ", dynamicRoutes=" + dynamicRoutes +
                ", requestHandler=" + requestHandler +
                '}';
    }
}

File: rust-lab/src/web_client/utils/web/route_tree.rs
use crate::web_client::enums::{RequestHandler, WILDCARD_STR};
use crate::web_client::utils::web::route_pattern::RoutePattern;
use std::collections::{HashMap, VecDeque};
use std::fmt::{Debug, Formatter};
use std::ops::Deref;

struct MatchCandidate<'a> {
    route_tree: &'a RouteTree,
    path_variables: HashMap<String, String>,
}
// #[derive(Debug)]
pub struct MatchResult<'a>{
    pub request_handler: &'a RequestHandler,
    pub path_variables: HashMap<String, String>,
}

impl Debug for MatchResult<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("MatchResult")
        .field("request_handler", &"request handler")
        .field("path_variables", &self.path_variables)
        .finish()
    }
}
pub struct RouteTree{
    segment: String,
    path: String,
    static_routes: HashMap<String, RouteTree>,
    dynamic_routes: HashMap<String, RouteTree>,
    request_handler: Option<RequestHandler>
}

impl RouteTree{
    pub fn new(segment: &str, path: &str) -> RouteTree{
        RouteTree{
            segment: segment.to_string(),
            static_routes: HashMap::new(),
            dynamic_routes: HashMap::new(),
            path: path.to_string(),
            request_handler: None
        }
    }
    pub fn add_route(&mut self, mut route: RoutePattern){
        let mut node = self;
        let segments = route.segments;
        for i in 0..segments.len(){
            let segment = &segments[i];
            let is_wildcard = WILDCARD_STR == segment.first;
            let next_level = if is_wildcard {&mut node.dynamic_routes} else { &mut node.static_routes };
            let _segment = if is_wildcard { segment.second.clone() } else { segment.first.clone() };
            if !next_level.contains_key(&_segment) {
                next_level.insert(_segment.clone(),RouteTree::new(&segment.first, &segment.second));
            }
            if i+1 == segments.len() {
                if let Some(handler) = route.handler.take(){
                    next_level.get_mut(&_segment).unwrap().request_handler = Some(handler);
                }
            }
            node = next_level.get_mut(&_segment).unwrap();
        }
    }
    pub fn find(&self, path: &str) -> Option<MatchResult> {
        let segments:Vec<&str> = path.split("/").collect();
        let mut queue:VecDeque<MatchCandidate> = VecDeque::new();
        queue.push_back(MatchCandidate{
            route_tree: self,
            path_variables: HashMap::new()
        });
        for segment in segments {
            let mut next_level:VecDeque<MatchCandidate> = VecDeque::new();
            while !queue.is_empty() {
                let candidate = queue.pop_front().unwrap();
                let child = candidate.route_tree.static_routes.get(&segment.to_string());
                if let Some(child) = child {
                    next_level.push_back(MatchCandidate{
                        route_tree: child,
                        path_variables: candidate.path_variables.clone(),
                    })
                }
                let children:Vec<&RouteTree> = candidate.route_tree.dynamic_routes.values().collect();
                if !children.is_empty() {
                    for child in children {
                        let mut path_variables = candidate.path_variables.clone();
                        path_variables.insert(child.path.clone(), segment.to_string());
                        next_level.push_back(MatchCandidate{
                            route_tree: child,
                            path_variables,
                        })
                    }
                }
            }
            queue = next_level;
        }
        while !queue.is_empty() {
            let candidate = queue.pop_front().unwrap();
            let handler = &candidate.route_tree.request_handler;
            if let Some(handler) = handler {
                let option = Some(MatchResult {
                    request_handler: handler,
                    path_variables: candidate.path_variables.clone(),
                });
                return option;
            }
        }
        None
    }
}

impl Debug for RouteTree{
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RouteTree")
        .field("segment", &self.segment)
        .field("children", &self.static_routes)
            .field("path", &self.path)
            .finish()
    }
}

#[cfg(test)]
mod test{
    use crate::web_client::health_check::health_check;
    use crate::web_client::utils::web::route_pattern::RoutePattern;
    use crate::web_client::utils::web::route_tree::RouteTree;

    #[test]
    pub fn match_path_and_variable(){
        let mut route_tree = RouteTree::new("", "");
        let pattern1 = RoutePattern::parse("/{pathVariable1}/{pathVariable2}/path", Box::new(health_check));
        // pattern1.handler = Some(Box::new(health_check));
        route_tree.add_route(pattern1);

        let pattern2 = RoutePattern::parse("/{pathVariable}/path", Box::new(health_check));
        // pattern2.handler = Some(Box::new(health_check));
        route_tree.add_route(pattern2);

        let pattern3 = RoutePattern::parse("/path/{pathVariable}", Box::new(health_check));
        // pattern3.handler = Some(Box::new(health_check));
        route_tree.add_route(pattern3);

        let pattern4 = RoutePattern::parse("/{pathVariable1}/path/{pathVariable2}/path", Box::new(health_check));
        // pattern4.handler = Some(Box::new(health_check));
        route_tree.add_route(pattern4);

        let option = route_tree.find("/1/2/path");
        assert!(option.is_some());
        let option = option.unwrap();
        println!("{:?}", option);

        let option = route_tree.find("/1/path");
        assert!(option.is_some());
        let option = option.unwrap();
        println!("{:?}", option);

        let option = route_tree.find("/path/1");
        assert!(option.is_some());
        let option = option.unwrap();
        println!("{:?}", option);

        let option = route_tree.find("/path/path/path/path");
        assert!(option.is_some());
        let option = option.unwrap();
        println!("{:?}", option);
        ()
    }
}